build-llvm.py

```diff
@@ -0,0 +1,402 @@
+    defines = {}
+    # Objective-C Automatic Reference Counting (we don't use Objective-C)
+    defines['CLANG_ENABLE_ARCMT'] = 'OFF'
+    # We don't (currently) use the static analyzer
+    defines['CLANG_ENABLE_STATIC_ANALYZER'] = 'OFF'
+    # We don't use the plugin system and this saves cycles according to Chromium OS
+    defines['CLANG_PLUGIN_SUPPORT'] = 'OFF'
+    # The C compiler to use
+    defines['CMAKE_C_COMPILER'] = cc
+    # The C++ compiler to use
+    defines['CMAKE_CXX_COMPILER'] = cxx
+    # Where the toolchain should be installed
+    defines['CMAKE_INSTALL_PREFIX'] = install_folder.as_posix()
+    # For LLVMgold.so, which is used for LTO with ld.gold
+    defines['LLVM_BINUTILS_INCDIR'] = root.joinpath(utils.current_binutils(),
+                                                    "include").as_posix()
+    # The projects to build
+    defines['LLVM_ENABLE_PROJECTS'] = projects
+    # Don't build bindings; they are for other languages that the kernel does not use
+    defines['LLVM_ENABLE_BINDINGS'] = 'OFF'
+    # Don't build Ocaml documentation
+    defines['LLVM_ENABLE_OCAMLDOC'] = 'OFF'
+    # Removes system dependency on terminfo and almost every major clang provider turns this off
+    defines['LLVM_ENABLE_TERMINFO'] = 'OFF'
+    # Don't build clang-tools-extras to cut down on build targets (about 400 files or so)
+    defines['LLVM_EXTERNAL_CLANG_TOOLS_EXTRA_SOURCE_DIR'] = ''
+    # Don't include documentation build targets because it is available on the web
+    defines['LLVM_INCLUDE_DOCS'] = ''
+    # Don't include example build targets to save on cmake cycles
+    defines['LLVM_INCLUDE_EXAMPLES'] = 'OFF'
+    # The architectures to build backends for
+    defines['LLVM_TARGETS_TO_BUILD'] = targets
+
+    # If a debug build was requested
+    if debug:
+        defines['CMAKE_BUILD_TYPE'] = 'Debug'
+        defines['CMAKE_C_FLAGS'] = '-march=native -mtune=native'
+        defines['CMAKE_CXX_FLAGS'] = '-march=native -mtune=native'
+        defines['LLVM_BUILD_TESTS'] = 'ON'
+    # If a release build was requested
+    else:
+        defines['CMAKE_BUILD_TYPE'] = 'Release'
+        defines['CMAKE_C_FLAGS'] = '-O2 -march=native -mtune=native'
+        defines['CMAKE_CXX_FLAGS'] = '-O2 -march=native -mtune=native'
+        defines['LLVM_INCLUDE_TESTS'] = 'OFF'
+        defines['LLVM_ENABLE_WARNINGS'] = 'OFF'
+
+    # Don't build libfuzzer when compiler-rt is enabled, it invokes cmake again and we don't use it
+    if "compiler-rt" in projects:
+        defines['COMPILER_RT_BUILD_LIBFUZZER'] = 'OFF'
+
+    # Use ccache if it is available for faster incremental builds
+    if shutil.which("ccache") is not None:
+        defines['LLVM_CCACHE_BUILD'] = 'ON'
+
+    # If we found a linker, we should use it
+    if ld is not None:
+        defines['LLVM_USE_LINKER'] = ld
+
+    # Add the defines, point them to our build folder, and invoke cmake
+    cmake = ['cmake', '-G', 'Ninja', '-Wno-dev']
```

Comment:
what's `-Wno-dev` for?

Reply:
I will occasionally see some cmake warnings that mention `-Wno-dev`. I don't want a user to think they are doing something wrong (since those warnings are more targeted at LLVM developers). I can remove it and see what happens.

Reply:
What do the warnings look like?

Reply:
```
CMake Warning (dev) at /usr/share/cmake-3.14/Modules/CheckIncludeFile.cmake:80 (message):
  Policy CMP0075 is not set: Include file check macros honor
  CMAKE_REQUIRED_LIBRARIES.  Run "cmake --help-policy CMP0075" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  CMAKE_REQUIRED_LIBRARIES is set to:

    m

  For compatibility with CMake 3.11 and below this check is ignoring it.
Call Stack (most recent call first):
  /home/nathan/cbl/git/tc-build/llvm-project/compiler-rt/cmake/base-config-ix.cmake:9 (check_include_file)
  /home/nathan/cbl/git/tc-build/llvm-project/compiler-rt/CMakeLists.txt:27 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Looking for unwind.h - found
-- Looking for rpc/xdr.h
-- Looking for rpc/xdr.h - not found
CMake Warning (dev) at /home/nathan/cbl/git/tc-build/llvm-project/compiler-rt/CMakeLists.txt:62 (set):
  implicitly converting 'BOOLEAN' to 'STRING' type.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at /home/nathan/cbl/git/tc-build/llvm-project/compiler-rt/CMakeLists.txt:65 (set):
  implicitly converting 'BOOLEAN' to 'STRING' type.
This warning is for project developers.  Use -Wno-dev to suppress it.
```

These are the ones I see.